### PR TITLE
Add loongarch64 asm! support

### PR DESCRIPTION
**Pull request by zhaixiaojuan**

### CODE REVIEW

**Comment by cjgillot** [comment]
As it's related to asm, @Amanieu could you take a look?

**Comment by Amanieu** [comment]
LGTM but this needs tests in src/test/assembly/asm.

**Comment by zhaixiaojuan** [comment]
src/test/assembly/asm/loongarch-type.rs  generates the corresponding assembly file:  [loongarch-type.s](https://github.com/rust-lang/rust/commit/8dbc47d60a26acc7841e8668e434fc0e10f56d1b)

**Comment by Amanieu** [comment]
It seems that CI is failing due to a check introduced by #77280.

cc @petrochenkov: The loongarch target was added in LLVM 15, and is not available on the PR CI which uses the system LLVM.

**Comment by Amanieu** [comment]
I think you might be able to fix this by adding `// min-llvm-version: 15.0` to that file.

**Comment by zhaixiaojuan** [comment]
> I think you might be able to fix this by adding `// min-llvm-version: 15.0` to that file.

Thanks a lot.

**Comment by bjorn3** [comment]
When COMPILETEST_NEEDS_ALL_LLVM_COMPONENTS is set, compiletest expects that all llvm components used by any test are available. This env var is set in CI. In other words you can't use make a test conditional on any llvm component that isn't enabled by default for rust it seems. If loongarch64 is already supported by the llvm version used by rust, you could add it to https://github.com/rust-lang/rust/blob/b32223fec10743af93913c41ec60645d741b0cb6/src/bootstrap/native.rs#L358 to enable it. (also make sure to add it to the `experimental_targets` field in `config.toml.example`).

**Comment by Amanieu** [comment]
I think you need to adjust this check: https://github.com/rust-lang/rust/blob/b32223fec10743af93913c41ec60645d741b0cb6/src/tools/compiletest/src/header.rs#L1037

Move the component detection after the LLVM version checking.

**Comment by zhaixiaojuan** [comment]
I adjusted the check order of llvm version and component in the file 
 src/tools/compiletest/src/header.rs.

File: src/test/assembly/asm/loongarch-type.rs [hunk 1]
```diff
@@ -0,0 +1,195 @@
+// min-llvm-version: 15.0
+// revisions: loongarch64
```
Comment (Amanieu) [inline src/test/assembly/asm/loongarch-type.rs:2]: This line is not needed.

File: src/tools/compiletest/src/header.rs [hunk 2]
```diff
@@ -1034,31 +1034,17 @@ fn ignore_llvm(config: &Config, line: &str) -> bool {
...
-            return true;
-        }
-    }
     if let Some(actual_version) = config.llvm_version {
         if let Some(rest) = line.strip_prefix("min-llvm-version:").map(str::trim) {
             let min_version = extract_llvm_version(rest).unwrap();
             // Ignore if actual version is smaller the minimum required
             // version
-            actual_version < min_version
+            return actual_version < min_version;
```
Comment (Amanieu) [inline src/tools/compiletest/src/header.rs:1042]:
```suggestion
            if actual_version < min_version {
                return true;
            }
```

File: src/tools/compiletest/src/header.rs [hunk 3]
```diff
@@ -1034,31 +1034,17 @@ fn ignore_llvm(config: &Config, line: &str) -> bool {
...
         } else if let Some(rest) = line.strip_prefix("min-system-llvm-version:").map(str::trim) {
             let min_version = extract_llvm_version(rest).unwrap();
             // Ignore if using system LLVM and actual version
             // is smaller the minimum required version
-            config.system_llvm && actual_version < min_version
+            return config.system_llvm && actual_version < min_version;
```
Comment (Amanieu) [inline src/tools/compiletest/src/header.rs:1047]: Same here.

File: src/tools/compiletest/src/header.rs [hunk 4]
```diff
@@ -1069,11 +1055,27 @@ fn ignore_llvm(config: &Config, line: &str) -> bool {
                 panic!("Malformed LLVM version range: max < min")
             }
             // Ignore if version lies inside of range.
-            actual_version >= v_min && actual_version <= v_max
+            return actual_version >= v_min && actual_version <= v_max;
```
Comment (Amanieu) [inline src/tools/compiletest/src/header.rs:1058]: Same here.

File: src/tools/compiletest/src/header.rs [hunk 5]
```diff
@@ -1069,11 +1055,27 @@ fn ignore_llvm(config: &Config, line: &str) -> bool {
...
         } else {
-            false
+            return false;
```
Comment (Amanieu) [inline src/tools/compiletest/src/header.rs:1060]: Remove the `else` branch: if there's no `min-llvm-version` then we should continue on to the next check below.

File: src/tools/compiletest/src/header.rs [hunk 6]
```diff
@@ -1069,11 +1055,27 @@ fn ignore_llvm(config: &Config, line: &str) -> bool {
...
+        if let Some(missing_component) = needed_components
+            .split_whitespace()
+            .find(|needed_component| !components.contains(needed_component))
+        {
+            if env::var_os("COMPILETEST_NEEDS_ALL_LLVM_COMPONENTS").is_some() {
+                panic!("missing LLVM component: {}", missing_component);
+            }
+            return true;
+        } else {
+            return false;
```
Comment (Amanieu) [inline src/tools/compiletest/src/header.rs:1076]:
```suggestion
            true
        } else {
            false
```
Use trailing return syntax.

File: src/tools/compiletest/src/header.rs [hunk 7]
```diff
@@ -1069,11 +1055,27 @@ fn ignore_llvm(config: &Config, line: &str) -> bool {
...
         }
     } else {
-        false
+        return false;
```
Comment (Amanieu) [inline src/tools/compiletest/src/header.rs:1079]:
```suggestion
        false
```

**Comment by zhaixiaojuan** [comment]
I seem to find a problem, when the iter_header function is called in make_test_description, the iter_header function loops to read the configuration information of each line in the xxx-types.rs file, and ignore_llvm is called every time during the loop. 
When reading min-llvm-revision, the min-llvm-version branch will be entered in the llvm-ignore function. But when reading the component, the version check will be skipped and go directly to the component branch, which results in a "missing LLVM component: loongarch" error.

make_test_description:
image
iter_header:
image


I'm not sure if my thinking is correct, my current understanding is:
Between lines 924 and 948 of the code, I think the logic here should be that as long as there is a configuration that makes the value of ignore true, then the value of ignore will be true at the end.
Therefore, when ignore is set to true in one loop of iter_header, the value of ignore does not need to be set in subsequent loops.So when setting the ignore value, I added the judgment condition.

[src/tools/compiletest/src/header.rs](https://github.com/rust-lang/rust/commit/f98a9fd90588f3881cd25f65b370b837163887c7#diff-bf53d26b9ca67283fc825e902a64780b463c153d1b1df0c8e782da0300f08956)
make_test_description:
```
923        if !ignore {
924            ignore |= ignore_llvm(config, ln);
 ......
948            ignore |= !has_rust_lld && config.parse_name_directive(ln, "needs-rust-lld");
949        }
```

**Comment by zhaixiaojuan** [comment]
The code in llvm-15 added support for LoongArch, but the LoongArch target is not built by default, causing it to fail in x86_64-gnu-stable tests.
Adding LoongArch to the experimental target in the src/bootstrap/native.rs file can build the target LoongArch in llvm, however the LoongArch target is not supported in the test environment of x86_64-gnu-llvm-13.
By communicating with the llvm community LoongArch maintainer SixWeining (Lu Weining), it is planned to add LoongArch to the llvm default build target in the llvm-16 version, so I currently change the min-llvm-revision to llvm 16.0.

**Comment by Amanieu** [comment]
On the full CI suite (run by bors) we compile our own LLVM. Have you tried enabling the target as suggested in https://github.com/rust-lang/rust/pull/101069#issuecomment-1237303232?

**Comment by zhaixiaojuan** [comment]
> On the full CI suite (run by bors) we compile our own LLVM. Have you tried enabling the target as suggested in [#101069 (comment)](https://github.com/rust-lang/rust/pull/101069#issuecomment-1237303232)?
>

I refer to this code, enable the LoongArch target, and test it locally. The "missing LLVM component: loongarch" was resolved during the test, but since the target loongarch64-unknown-linux-gnuf64 is not supported in the currently used rustc, the following error will be reported:
```
--- stderr -------------------------------
error: Error loading target specification: Could not find specification for target "loongarch64-unknown-linux-gnuf64". Run `rustc --print target-list` for a list of built-in targets
------------------------------------------
```

Is it possible to consider "//ignore-test" in the file loongarch-type.rs to ignore tests, or other suggestions and methods.

**Comment by Amanieu** [comment]
Ah I see the loongarch target is added in #96971 which is not merged yet. I think we should delay this PR until #96971 is merged.

File: compiler/rustc_target/src/asm/loongarch.rs [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        #error = ["$r2","$tp"] =>
+            "reserved for TLS",
+        #error = ["$r3","$sp"] =>
+            "the stack pointer cannot be used as an operand for inline asm",
+        #error = ["$r21"] =>
+            "reserved by the ABI",
+        #error = ["$r22","$fp"] =>
+            "the frame pointer cannot be used as an operand for inline asm",
+        #error = ["$r31","$s8"] =>
+            "$r31 is used internally by LLVM and cannot be used as an operand for inline asm",
```
Comment (xen0n) [inline compiler/rustc_target/src/asm/loongarch.rs:118]: Why is this the case? I haven't seen any such usage in upstream LLVM so far. Care sharing a link?

File: compiler/rustc_target/src/asm/loongarch.rs [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+        f31: freg = ["$f31","$fs7"],
+        #error = ["$r0","$zero"] =>
+            "constant zero cannot be used as an operand for inline asm",
+        #error = ["$r2","$tp"] =>
+            "reserved for TLS",
+        #error = ["$r3","$sp"] =>
+            "the stack pointer cannot be used as an operand for inline asm",
+        #error = ["$r21"] =>
+            "reserved by the ABI",
+        #error = ["$r22","$fp"] =>
```
Comment (xen0n) [inline compiler/rustc_target/src/asm/loongarch.rs:115]: `$fp` is also `$s9`, please add this ABI name if possible.

**Review by xen0n — COMMENTED** [review]
Please delay until #96971 is merged, as others have pointed out.

File: compiler/rustc_target/src/asm/loongarch.rs (hunk 8 above)
Comment (Amanieu) [reply compiler/rustc_target/src/asm/loongarch.rs:118]: https://github.com/llvm/llvm-project/blob/bd81524e7f52c38ad0bc689934343a476e545265/llvm/lib/Target/LoongArch/LoongArchRegisterInfo.cpp#L93
https://github.com/llvm/llvm-project/blob/77c19773843dbdf9ab7f3415b001fa1475ef9fa4/llvm/lib/Target/LoongArch/MCTargetDesc/LoongArchBaseInfo.cpp#L36

**Comment by apiraino** [comment]
Based [on this comment](https://github.com/rust-lang/rust/pull/101069#issuecomment-1252337831) I'll switch to waiting on author and try to signal that this pr should wait a little bit (hope I am reading it correctly). Feel free to request a review with `@rustbot ready` when work here can progress again. thanks!

@rustbot author

**Comment by Amanieu** [comment]
I'm changing the status to `S-blocked` since it's blocked waiting on #96971 being merged.

**Comment by heiher** [comment]
@Amanieu I recently updated the patch set and noticed that one of the CI tests failed. It appears that the failling test is bound to an older version that does not support LoongArch. I would greatly appreciate it if you could provide me with some guidance on how to resolve this issue.

**Comment by heiher** [comment]
> @Amanieu I recently updated the patch set and noticed that one of the CI tests failed. It appears that the failling test is bound to an older version that does not support LoongArch. I would greatly appreciate it if you could provide me with some guidance on how to resolve this issue.

It seems we should check the version before checking the components to ignore tests that do not match the `min-llvm-version`.

**Comment by Amanieu** [comment]
Should be fixed by #110232.

**Review by Amanieu — COMMENTED** [review]
Can you also update the documentation in the unstable book for `src/doc/unstable-book/src/language-features/asm-experimental-arch.md`.

**Comment by lqd** [comment]
cranelift-codegen / keccak are currently noisy, and this PR looks completely unrelated, @rustbot label: +perf-regression-triaged.

**Pull request merged**